File: solutions/Leetcode_954/leetcode_954.py
from collections import Counter
from typing import List

class Solution1:
    def can_reorder_doubled(self, arr: List[int]) -> bool:
        num_count = Counter(arr)
        for x in sorted(num_count.keys(), key=abs):
            if num_count[x] > num_count[2*x]:
                return False
            num_count[2*x] -= num_count[x]
        return True

File: solutions/Leetcode_954/leetcode_954.cpp
#include <algorithm>
#include <unordered_map>
#include <utility>
#include <vector>

class Solution1
{
public:
    bool canReorderDoubled(std::vector<int> &arr)
    {
        std::unordered_map<int, int> numCount;
        for (const int x : arr)
            numCount[x]++;
        std::vector<int> keys;
        for (std::pair<const int, int> item : numCount)
            keys.push_back(item.first);
        std::sort(keys.begin(), keys.end(), [](int x, int y){ return abs(x) < abs(y); });
        for (const int x : keys)
        {
            if (numCount[x] > numCount[x * 2])
                return false;
            numCount[x * 2] -= numCount[x];
        }
        return true;
    }
};